### PR TITLE
Feature/mapped types

### PR DESCRIPTION
**Pull request by mweststrate**
Implements #705

### CODE REVIEW

**Comment by ryedin** [comment]
Will this fix #782 that I just added?

**Comment by mweststrate** [comment]
First version can now be tried! `mobx-state-tree@2.0.4-types.1`

File: README.md [hunk 1]
```diff
@@ -1156,10 +1156,12 @@ Yes, with MST it is pretty straight forward to setup hot reloading for your stor
 
 ### TypeScript & MST
 
-TypeScript support is best-effort, as not all patterns can be expressed in TypeScript. But except for assigning snapshots to properties we get pretty close! As MST uses the latest fancy Typescript features it is recommended to use TypeScript 2.3 or higher, with `noImplicitThis` and `strictNullChecks` enabled.
+TypeScript support is best-effort, as not all patterns can be expressed in TypeScript. But except for assigning snapshots to properties we get pretty close! As MST uses the latest fancy Typescript features it is required to use TypeScript 2.8 or higher, with `noImplicitThis` and `strictNullChecks` enabled.
```
Comment (Bnaya) [inline README.md]: Dose applySnapshot is typesafe?

**Comment by FredyC** [comment]
I am bit confused here. Is this change included in latest version 2.0.5? I mean it feels weird to _downgrade_ to 2.0.4-types.1 to get something new :)

**Comment by AndrewSorokin** [comment]
2.0.4-types.1 has the same issue as https://github.com/mobxjs/mobx-state-tree/issues/703

**Comment by mweststrate** [comment]
no it isn't, but it was forked from 2.0.4, so that is why. Will probably
merge this later this week after some more testing

Op di 22 mei 2018 om 10:02 schreef AndrewSorokin <notifications@github.com>:

**Comment by skellock** [comment]
FWIW @mweststrate I’ve experienced a pretty big slow down on the new types branch in my current app. 

VS Code crawls when working with our store models and even a fresh tsc is twice as slow.

I’d love to upgrade this feedback from near-useless to moderately-helpful but I’m not quite sure how to benchmark typescript.

**Comment by Bnaya** [comment]
@skellock look at https://github.com/mobxjs/mobx-state-tree/issues/705#issuecomment-382152973
from that specific comment and on, there are several tips, but they add some overhead.

**Pull request merged**